<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
  <title>Affymetrix CDF File Format</title>
  <meta http-equiv="Content-Type"
 content="text/html; charset=windows-1252">
  <meta
 content="Affymetrix, GeneChip, GeneChip Array, Netaffx, NetAffx Analysis Center, Affymetrics, Affimetrix, Affimetrics, "
 name="keywords">
  <meta
 content="Affymetrix is dedicated to developing state-of-the-art technology for acquiring, analyzing, and managing complex genetic information for use in biomedical research. Affymetrix sells GeneChip® brand arrays."
 name="description">
  <script language="JavaScript">
  </script>
  <script language="JavaScript" src="files_files/main.js"
 type="text/javascript"></script>
  <link href="files_files/style.css" type="text/css" rel="stylesheet">
  <link href="/favicon.ico" rel="icon">
  <meta content="Microsoft FrontPage 6.0" name="GENERATOR">
  <style>
<!--
table.MsoTableClassic4
	{border-top:1.5pt solid black;
	border-left:1.0pt solid black;
	border-bottom:1.5pt solid black;
	border-right:1.0pt solid black;
	font-size:10.0pt;
	font-family:"Times New Roman";
	}
-->
  </style>
</head>
<body onresize="fixIt()" leftmargin="0" topmargin="0"
 onunload="window.focus()" alink="#003366" bgcolor="#ffffff"
 link="#003366" marginheight="0" marginwidth="0" vlink="#003366">
<!-- Top links and Search plus noindex tags for atomz--> <noindex> </noindex>
<table border="0" cellpadding="0" cellspacing="0">
  <tbody>
    <tr>
      <td align="left" valign="center">&nbsp;</td>
    </tr>
  </tbody>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="601">
  <tbody>
    <tr>
      <td align="left" valign="top"><img alt="Affymetrix"
 src="files_files/affymetrix_logo.gif" border="0" height="54"
 width="128"></td>
      <td><br>
      </td>
      <td class="pageheader" align="left" valign="center">
      <p>AFFYMETRIX DEVELOPER NETWORK</p>
      </td>
    </tr>
  </tbody>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="772">
  <tbody>
    <tr>
      <td align="left" bgcolor="#336699" height="2" valign="top"
 width="772"><img alt="" src="files_files/spacer.gif" border="0"
 height="3" width="1"></td>
    </tr>
  </tbody>
</table>
<table background="files_files/network_background.gif" border="0"
 cellpadding="5" cellspacing="0" width="409">
  <tbody>
    <tr>
      <td align="left" valign="top">
      <table background="files_files/spacer.gif" border="0"
 cellpadding="0" cellspacing="0">
        <tbody>
          <tr>
            <td align="left" valign="top" width="11"><img alt=""
 src="files_files/spacer.gif" border="0" height="1" width="11"></td>
            <td align="left" valign="top" width="100%"><img alt=""
 src="files_files/spacer.gif" border="0" height="41" width="1"></td>
          </tr>
          <tr>
            <td align="left" valign="top" width="11"><img alt=""
 src="files_files/spacer.gif" border="0" height="1" width="11"></td>
            <td class="intro" align="left" valign="top">
            <p><span class="header2">Affymetrix® CDF Data File Format</span></p>
            </td>
          </tr>
        </tbody>
      </table>
      </td>
    </tr>
  </tbody>
</table>
<br>
<table border="0" cellpadding="5" cellspacing="0">
  <tbody>
    <tr>
      <td align="left" valign="top">
      <table border="0" cellpadding="0" cellspacing="0">
        <tbody>
          <tr>
            <td align="left" valign="top">
            <table border="0" cellpadding="0" cellspacing="0">
              <tbody>
                <tr>
                  <td style="border-style: none;" align="right"
 valign="top" width="11"><img alt="" src="files_files/bullet2.gif"
 border="0" height="11" width="11"></td>
                  <td style="border-style: none;" align="left"
 valign="top" width="8"><img alt="" src="files_files/spacer.gif"
 border="0" height="1" width="5"></td>
                  <td class="header3" style="border-style: none;"
 align="left" valign="center"><a name="CDF_FILE"></a>CDF FILE</td>
                </tr>
                <tr>
                  <td colspan="3" style="border-style: none;"
 align="left" valign="top"><img alt="" src="files_files/spacer.gif"
 border="0" height="10" width="1"></td>
                </tr>
                <tr>
                  <td style="border-style: none;" align="left"
 valign="top" width="11"><img alt="" src="files_files/spacer.gif"
 border="0" height="1" width="11"></td>
                  <td style="border-style: none;" align="left"
 valign="top" width="8">
                  <p><img alt="" src="files_files/spacer.gif" border="0"
 height="1" width="5"></p>
                  <p>&nbsp;</p>
                  </td>
                  <td class="content" style="border-style: none;"
 align="left" valign="top">
                  <p><span class="header4">Description</span><br>
The CDF file describes the layout for an Affymetrix GeneChip array. An
array may contain Expression, Genotyping, CustomSeq, Copy Number and/or
Tag probe sets. All probe set names within an array are unique.
Multiple copies of a probe set may exist on a single array as long as
each copy has a unique name.</p>
                  <p>The information below will describe the following
versions:</p>
                  <ul>
                    <li> <a href="#CDF_TEXT">ASCII text format</a> is
used by the MAS and GCOS 1.0 software. This was also known as the ASCII
version.</li>
                    <li> <a href="#CDF_XDA">XDA format</a> is used by
the GCOS 1.2 and above software. This was also know as the binary or
XDA version.</li>
                  </ul>
                  <p><span class="header4"><a name="CDF_TEXT"></a>ASCII
Text Format</span><br>
The format of this CDF file is an ASCII text file similar to the
Windows INI format.</p>
                  <p>The file is divided up into sections. The start of
each section is defined by a line containing a section name enclosed in
square braces. The section names are: "CDF", "Chip", "QC<i>I</i>"
(where I ranges from 1 to the number of QC probe sets), "Unit<i>J</i>"
(where J is an internal index to uniquely distinguish probe
sets),&nbsp; and "Unit<i>J</i>_Block<i>K</i>" (where J and K are
internal indicies used to distinguish subsets of a probe set). The data
in each section is of the format TAG=VALUE.</p>
                  <p>The "CDF" section contains the version number of
the file. The TAGS are:</p>
                  <table span="" class="content" border="1"
 cellpadding="1" cellspacing="0">
                    <tbody>
                      <tr>
                        <td height="14" width="110"><b>TAG</b></td>
                        <td height="14" width="70%"><b>Description</b></td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Version</td>
                        <td height="14" width="70%">The version number.
Should always be set to "GC1.0", "GC2.0", "GC3.0", "GC4.0" or "GC5.0".
This
document describes GC3.0, GC4.0, and GC5.0 version CDF files.</td>
                      </tr>
                    </tbody>
                  </table>
                  <p>The "Chip" section contains the following TAGS:</p>
                  <table span="" class="content" border="1"
 cellpadding="1" cellspacing="0">
                    <tbody>
                      <tr>
                        <td height="14" width="110"><b>TAG</b></td>
                        <td height="14" width="70%"><b>Description</b></td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Name</td>
                        <td height="14" width="70%">The name of the
array. This item is not used by the software.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Rows</td>
                        <td height="14" width="70%">The number of rows
of cells on the array.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Cols</td>
                        <td height="14" width="70%">The number of
columns of cells on the array.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumberOfUnits</td>
                        <td height="14" width="70%">The number of units
in the array not including QC units. For CustomSeq arrays, there are
2 units: Unit1 contains the probes interrogating a sense target and
Unit2 contains the probes interrogating an anti-sense target. For all
other array types, there exists one unit per probe set.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> MaxUnit</td>
                        <td height="14" width="70%">Each unit is given
a unique number. This value is the maximum of the unit numbers of all
the units in the array (not including QC units).</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumQCUnits</td>
                        <td height="14" width="70%">The number of QC
units. QC units are defined in version 2 and above. CustomSeq arrays do
not contain any QC units.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> ChipReference</td>
                        <td height="14" width="70%">Used for CustomSeq,
HIV and P53 arrays only. This is the reference sequence displayed by
the Affymetrix software. The sequence may contain spaces. This value is
defined for version 2 and above.</td>
                      </tr>
                    </tbody>
                  </table>
                  <p>The next set of sections where the name begins
with "QC" define the QC units or probe sets in the array. There are
NumQCUnits (from the Chip section) QC sections. Each section name is a
combination
of "QC" and an index ranging from 1 to NumQCUnits-1 and will be listed
sequentially.
QC
units
are
defined
for
version 2 and above. Each section contains the following TAGS:</p>
                  <table span="" class="content" border="1"
 cellpadding="1" cellspacing="0">
                    <tbody>
                      <tr>
                        <td height="14" width="110"><b>TAG</b></td>
                        <td height="14" width="70%"><b>Description</b></td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Type</td>
                        <td height="14" width="70%">Defines the type of
QC probe set. The defined types are:
                        <p>0 - Unknown<br>
1 - Checkerboard Negative<br>
2 - Checkerboard Positive<br>
3 - Hybridization Negative<br>
4 - Hybridization Positive<br>
5 - Text Features Negative<br>
6 - Text Features Positive<br>
7 - Central Negative <br>
8 - Central Positive<br>
9 - Gene Expression Negative<br>
10 - Gene Expression Positive<br>
11 - Cycle Fidelity Negative<br>
12 - Cycle Fidelity Positive<br>
13 - Central Cross Negative<br>
14 - Central Cross Positive<br>
15 – Cross Hyb Negative<br>
16 – Cross Hyb Positive</p>
                        </td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumberCells</td>
                        <td height="14" width="70%">The number of cells
in the probe set.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> CellHeader</td>
                        <td height="14" width="70%">Defines the data
contained in the subsequent lines, separated by tabs.
                        <p>For all QC probe set types:<br>
                        <strong>X</strong> - The X coordinate of the
cell.<br>
                        <strong>Y</strong> - The Y coordinate of the
cell.<br>
                        <strong>PROBE</strong> - The probe sequence of
the cell. Typically set to "N".<br>
                        <strong>PLEN</strong> - The number of bases in
the probe sequence.<br>
                        <strong>ATOM</strong> - An index used to group
multiple cells.<br>
                        <strong>INDEX</strong> - An index used to look
up the corresponding cell data in the CEL file.</p>
                        <p>The final data items are dependent on the
type of the QC probe set:<br>
                        <strong>MATCH</strong> - A boolean flag
indicating a perfect match probe. For types: 7 - Central Negative, 8 -
Central Positive, 9 - Gene Expression Negative,
10 - Gene Expression Positive<br>
                        <strong>BG</strong> - A boolean flag indicating
a background (blank) cell. For types: 9 - Gene Expression Negative, 10
- Gene Expression Positive<br>
                        <strong>CYCLES</strong> - This item is always a
list of 0's separated by a tab. There are as many 0's as number of
bases in the probe sequence (PLEN). For types: 11 - Cycle Fidelity
Negative, 12 - Cycle
Fidelity Positive</p>
                        </td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Cell<i>i</i></td>
                        <td height="14" width="70%">This contains the
information about a cell that belongs to the probe set. The value of <i>i</i>
in the tag ranges from 1 to the number of cells in the probe set
and will be listed sequentially. The values in each line depend on the
CellHeader.
The values are separated by
tabs.</td>
                      </tr>
                    </tbody>
                  </table>
                  <p>The next set of sections where the name begins
with "Unit" define the probes that are a member of the unit (probe
set). Each unit
is
divided into subsections termed "Blocks" which are referred to as
"groups" in the Files SDK documentation. </p>
                  <p>Each section name is a combination of "Unit" and
an index. There is no meaning to the index value. Immediately following
the "Unit" section there will be the "Block" sections for that unit
before the next unit is defined. </p>
                  <p>Each "Unit" section contains the following TAGS:</p>
                  <table span="" class="content" border="1"
 cellpadding="1" cellspacing="0">
                    <tbody>
                      <tr>
                        <td height="14" width="110"><b>TAG</b></td>
                        <td height="14" width="70%"><b>Description</b></td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Name</td>
                        <td height="14" width="70%">The name of the
unit. The probe set name for Genotyping, Copy Number, Polymorphic
Marker and Multichannel Marker units or "NONE" for all other unit types.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Direction</td>
                        <td height="14" width="70%">Defines if the
probes are interrogating a sense target or anti-sense target (1 -
sense, 2 - anti-sense, 3 - both).</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumAtoms</td>
                        <td height="14" width="70%">The number of atoms
in the entire probe set. This TAG name contain two values after the
equal sign. The first is the number of atoms and the second (if found)
is the number of cells
in each atom. An atom is a probe quartet for CustomSeq units and a
probe pair for all other unit types. </td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumCells</td>
                        <td height="14" width="70%">The number of cells
in the entire probe set. Probe pairs contain 2 cells and probe quartets
contain 4 cells.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> UnitNumber</td>
                        <td height="14" width="70%">An arbitrary index
value for the probe set.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> UnitType</td>
                        <td height="14" width="70%">Defines the type of
unit (0 - Unknown, 1 - CustomSeq, 2 - Genotyping, 3 - Expression, 7 -
Tag/GenFlex, 8 - Copy Number, 9 - Genotyping Control, 10 - Expression
Control, 11 - Polymorphic Marker, 12 - Multichannel Marker). An array
may
contain units of
varying types.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumberBlocks</td>
                        <td height="14" width="70%">The number of
blocks or groups in the probe set. </td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> MutationType</td>
                        <td height="14" width="70%">Used for Genotyping
units only in defining the type of polymorphism (0 - substitution, 1 -
insertion, 2 - deletion). This value is available in version 2 and
above.</td>
                      </tr>
                    </tbody>
                  </table>
                  <p>After the "Unit" section follows the&nbsp;
"Unit_Block" sections. There are as many "Unit_Block" sections as
defined
by NumberBlocks. A block will list the probes as its members.</p>
                  <p>The TAGS are:</p>
                  <table span="" class="content" border="1"
 cellpadding="1" cellspacing="0">
                    <tbody>
                      <tr>
                        <td height="14" width="110"><b>TAG</b></td>
                        <td height="14" width="70%"><b>Description</b></td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Name</td>
                        <td height="14" width="70%">The name of the
block. For Genotyping units this is the allele. For Polymorphic Marker
and Multichannel
Marker units this is "None". For all other unit types this is the name
of the
probe set. </td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> BlockNumber</td>
                        <td height="14" width="70%">An index to the
block.</td>
                      </tr>
                      <tr>
                        <td style="vertical-align: top;">Wobble<br>
                        </td>
                        <td style="vertical-align: top;">The wobble
situation for Polymorphic Marker and Multichannel Marker units in the
block.
Only available in
version 4 and 5.<br>
                        </td>
                      </tr>
                      <tr>
                        <td style="vertical-align: top;">Allele<br>
                        </td>
                        <td style="vertical-align: top;">The allele
code for Polymorphic Marker and Multichannel Marker units in the block.
Only
available in
version 4 and 5.</td>
                      </tr>
                      <tr>
                        <td style="vertical-align: top;">Channel<br>
                        </td>
                        <td style="vertical-align: top;">The channel
code for multichannel microarray platform. Only available in version 5.<br>
                        </td>
                      </tr>
                      <tr>
                        <td style="vertical-align: top;">RepType<br>
                        </td>
                        <td style="vertical-align: top;">The probe
replication type (0 - unknown, 1 - different probe sequences, 2 - some
probe sequences are identical, 3 - all probe sequences are identical)
for probe set groups used under multichannel microarray platform. Only
available in version 5.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumAtoms</td>
                        <td height="14" width="70%">The number of atoms
in the block.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> NumCells</td>
                        <td height="14" width="70%">The number of cells
in the block.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> StartPosition</td>
                        <td height="14" width="70%">The position of the
first atom.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> StopPosition</td>
                        <td height="14" width="70%">The position of the
last atom.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Direction</td>
                        <td height="14" width="70%">Used for
Genotyping, Polymorphic Marker
and Multichannel
Marker units only in defining whether the probes are interrogating a
sense
target or anti-sense target (0 - no direction, 1 - sense, 2 -
anti-sense). This value is available in version 3 and above.</td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> CellHeader</td>
                        <td height="14" width="70%">Defines the data
contained in the subsequent lines, separated by tabs. The values are:
                        <p><strong>X</strong>- The X coordinate of the
cell.<br>
                        <strong>Y</strong> - The Y coordinate of the
cell.<br>
                        <strong>PROBE</strong>- The probe sequence of
the cell. Typically set to "N".<br>
                        <strong>FEAT</strong> - Unused string.<br>
                        <strong>QUAL</strong> - The probe set name plus
the allele for Genotyping units. The probe set name for all other unit
types.<br>
                        <strong>EXPOS</strong> - Ranges from 0 to the
NumAtoms - 1 for Expression units. For all other unit types, provides
relative positional information for the probe.<br>
                        <strong>PLEN</strong> - The length of probe
sequence. Only available in version 4 and 5.<br>
                        <strong></strong><strong>POS</strong> - An
index to the base position within the probe where the mismatch occurs.<br>
                        <strong>CBASE</strong> - Not used.<br>
                        <strong>PBASE</strong> - The probe base at the
substitution position.<br>
                        <strong>TBASE</strong> - The base of the target
where the probe interrogates at the substitution position.<br>
                        <strong>ATOM</strong> - An index used to group
probe pairs or quartets. For Expression, identical to EXPOS.<br>
                        <strong>INDEX</strong> - An index used to look
up the corresponding cell data in the CEL file.<br>
                        <strong>GROUP</strong> - The physical grouping
of probe on the array. Only available in version 4 and 5.<br>
                        </p>
                        <p>The following are only available in version
2 and above:<br>
                        <strong>CODONIND</strong> - Always set to -1<br>
                        <strong>CODON</strong> -Always set to -1<br>
                        <strong>REGIONTYPE</strong> - Always set to 99<br>
                        <strong>REGION</strong> - Always set to a blank
character</p>
                        </td>
                      </tr>
                      <tr>
                        <td height="14" width="110"> Cell<i>i</i></td>
                        <td height="14" width="70%">This contains the
information about a cell that belongs to the block. The value of <i>i</i>
in the tag ranges from 1 to the number of cells in the block. The
values in each line depend on the CellHeader. The values are separated
by tabs.</td>
                      </tr>
                    </tbody>
                  </table>
                  <p><span class="header4"><a name="CDF_XDA"></a> XDA
Format</span><br>
The format of this CDF file is an binary file created for faster access
and smaller file size. The values in the file are stored in
little-endian format.</p>
                  <p>The file contents are define by:</p>
                  <p>
                  <table span="" class="content" border="1"
 cellpadding="1" cellspacing="0" height="1">
                    <tbody>
                      <tr>
                        <td height="12" width="110"><b>Item</b></td>
                        <td height="12" width="70%"><b>Description</b></td>
                        <td height="12" width="214"><b>Type</b></td>
                      </tr>
                      <tr>
                        <td height="14" width="110">1</td>
                        <td height="14" width="70%">Magic number.
Always set to 67.</td>
                        <td height="14" width="214">integer</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">2</td>
                        <td height="14" width="70%">Version number.
Should set to 1, 2, or 3.<br>
                        </td>
                        <td height="14" width="214">integer</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">3</td>
                        <td height="14" width="70%">The number of
columns of cells on the array.</td>
                        <td height="14" width="214">unsigned short</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">4</td>
                        <td height="14" width="70%">The number of rows
of cells on the array.</td>
                        <td height="14" width="214">unsigned short</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">5</td>
                        <td height="14" width="70%">The number of units
in the array not including QC units. The term unit is an internal term
which means probe set.</td>
                        <td height="14" width="214">integer</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">6</td>
                        <td height="14" width="70%">The number of QC
units.</td>
                        <td height="14" width="214">integer</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">7</td>
                        <td height="14" width="70%">The length of the
CustomSeq reference sequence.</td>
                        <td height="14" width="214">integer</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">8</td>
                        <td height="14" width="70%">The CustomSeq
reference sequence.</td>
                        <td height="14" width="214">char[ length
defined above]</td>
                      </tr>
                      <tr>
                        <td height="12" width="110">9</td>
                        <td height="12" width="70%">The probe set name.
The UNIT name for CustomSeq, Genotyping, Polymorphic Marker, and
Multichannel Marker. The BLOCK name for
Expression.</td>
                        <td height="12" width="214">char[64] * (# of
units)</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">10</td>
                        <td height="14" width="70%">File position for
the start of each QC unit information block.</td>
                        <td height="14" width="214">integer * (# of QC
units)</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">11</td>
                        <td height="14" width="70%">File position for
the start of each unit information block.</td>
                        <td height="14" width="214">integer * (# of
units)</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">12</td>
                        <td height="14" width="70%">QC information,
repeated for each QC unit:
                        <p> Type - unsigned short<br>
Number of probes - integer</p>
                        <p>Probe information, repeated for each probe
in the QC unit:</p>
X coordinate - unsigned short<br>
Y coordinate - unsigned short<br>
Probe length - unsigned char<br>
Perfect match flag - unsigned char<br>
Background probe flag - unsigned char<br>
                        </td>
                        <td height="14" width="214">see description</td>
                      </tr>
                      <tr>
                        <td height="14" width="110">13</td>
                        <td height="14" width="70%">Unit information,
repeated for each unit:
                        <p>UnitType - unsigned short (1 - Expression, 2
- Genotyping, 3 - CustomSeq, 4 - Tag, 5 - Copy Number, 6 - Genotyping
Control, 7 - Expression Control, 8 - Polymorphic Marker, 9 -
Multichannel Marker)<br>
Direction - unsigned char<br>
Number of atoms - integer<br>
Number of blocks - integer (always 1 for Expression units)<br>
Number of cells - integer<br>
Unit number (probe set number) - integer<br>
Number of cells per atom - unsigned char</p>
                        <p>Block information, repeated for each block
in the unit:</p>
                        <p>Number of atoms - integer<br>
Number of cells - integer<br>
Number of cells per atom - unsigned char<br>
Direction - unsigned char<br>
The position of the first atom - integer<br>
&lt;unused integer value&gt; - integer<br>
The block name - char[64]<br>
Wobble situation - unsigned short (only available in version 2 and 3)<br>
Allele code - unsigned short (only available in version 2 and 3)<br>
Channel - unsigned char (only available in version 3)<br>
RepType - unsigned char (0 - unknown, 1 - different probe sequences,
2 - some probe sequences are identical, 3 - all probe sequences are
identical) (Only available in version 3) </p>
                        <p>Cell information, repeated for each cell in
the block:</p>
                        <p>Atom number - integer<br>
X coordinate - unsigned short<br>
Y coordinate - unsigned short<br>
Index position (relative to sequence for CustomSeq, Genotyping, Copy
Number, Polymorphic Marker, and Multichannel Marker units, for
Expression
units this value is the atom number)
- integer<br>
Base of probe at substitution position - char<br>
Base of target at interrogation position - char<br>
Length of probe sequence - unsigned short (only available in version 2
and 3)<br>
Physical grouping of probe - unsigned short (only available in version
2 and 3)<br>
                        </p>
                        </td>
                        <td height="14" width="214">see description</td>
                      </tr>
                    </tbody>
                  </table>
                  </p>
                  </td>
                </tr>
              </tbody>
            </table>
            </td>
          </tr>
        </tbody>
      </table>
      </td>
    </tr>
  </tbody>
</table>
<p><br>
</p>
</body>
</html>
